amazon/ion/reader_binary.py

```diff
@@ -208,406 +208,275 @@ def _parse_sid_iter(data):
+
+    if not event:
+        return event, buffer
+    else:
+        return event.derive_field_name(SymbolToken(None, field_sid)), buffer
+
+
+class _ContextFrame(NamedTuple):
+    parser: Callable[[_ParserContext], Tuple[IonEvent, SliceableBuffer]]
+    ion_type: Optional[IonType]
+    depth: int
+    limit: int
+
+
+@coroutine
+def stream_handler():
+    """
+    Handler for an Ion Binary value-stream.
+    """
+    buffer = SliceableBuffer.empty()
+    # top-level context limit is -1 to denote no limit
+    context_stack = deque([_ContextFrame(_tlv_parser, None, 0, -1)])
+    cursor = 0
+    ion_event = None
+    skip_or_next = ReadEventType.NEXT
+    expect_data = False
+    # will get swapped out for tlv or struct parser in main loop
+    parser = _ivm_parser
+    parent_type = None
+    depth = 0
+    limit = -1
+
+    # This is the main event loop for the parser coroutine.
+    #
+    # Each iteration begins with responding with previous ion_event (None to start)
+    # and receiving the user's read event.
+    #
+    # Then there are two main parts:
+    # 1/ handles the users' request, checking invariants and extending the
+    #    buffer and/or skipping.
+    # 2/ parses if possible then mutates state based on the results of parsing.
+    #
+    # You should think of them as distinct functions, inlined to avoid stack
+    # push/pop overhead.
     while True:
-        data_event, self = (yield transition)
-        if data_event is not None and data_event.type is ReadEventType.SKIP:
-            yield ctx.read_data_transition(ctx.remaining, self, skip=True)
-
-        if ctx.queue.position == ctx.limit:
-            # We are at the end of the container.
-            # Yield the close event and go to enclosing container.
-            yield Transition(
-                IonEvent(IonEventType.CONTAINER_END, ion_type, depth=ctx.depth-1),
-                ctx.whence
-            )
-
-        if ion_type is IonType.STRUCT:
-            # Read the field name.
-            self_handler = _create_delegate_handler(self)
-            (field_sid, _), _ = yield ctx.immediate_transition(
-                _var_uint_field_handler(self_handler, ctx)
-            )
-            field_name = SymbolToken(None, field_sid)
+        read_event = yield ion_event
+        assert read_event is not None
+
+        # part 1: handle user's read event
+        if expect_data:
+            if read_event.type is not ReadEventType.DATA:
+                raise TypeError("Data expected!")
+            buffer = buffer.extend(read_event.data)
         else:
-            field_name = None
+            if read_event.type is ReadEventType.DATA:
+                raise TypeError("Next or Skip expected!")
+            skip_or_next = read_event.type
+
+        ion_event = None
+        if skip_or_next is ReadEventType.SKIP:
+            if parent_type is None:
+                raise TypeError("Cannot Skip Top-level values!")
```

Comment:
This message doesn't look quite right. You should be able to skip top-level values (e.g. a container at the top level). Is the problem that skip is being requested before `NEXT` to position the reader on the value?

Reply:
Yea... I added the message only based on what the `reader_trampoline` says.